hack/generator/pkg/astmodel/types.go

```diff
@@ -54,6 +54,24 @@ func (types Types) AddTypes(otherTypes Types) {
 	}
 }
 
+// AddWithEqualityCheck attempts to add the specified definition to the types collection.
+// Multiple adds of a type with the same shape are allowed, but attempting to add two
+// types with the same name but different shape will trigger an error.
+func (types Types) AddWithEqualityCheck(def TypeDefinition) error {
```

Comment:
Does this need to be a separate method? I'm thinking that it could be merged into the existing `Add()` method as it really is safe to add something that's already there.

Reply:
Aren't they expressing two different intents?

`Add` as it exists today panic's if there are duplicates. This is sorta nice for cases where you really don't expect to ever have duplicates (which is a lot of the time when you're say processing every type in the `Types` collection once, which a lot of our pipeline stages do).

`AddWithEqualityCheck` lets you expand what you're willing to tolerate (some duplicates, but only if they are structurally identical).

A lot of the common cases where we call `Add` today we really don't expect any duplicates at all because we're effectively doing a 1-1 mapping from an old `Types` set to a new `Types` set. It's only when you start treating things as a graph you run into duplicates and end up needing this.

I'd be fine combining them though if folks felt we don't really care about the above subtlety

Reply:
🤔 

I was thinking about it from an outcome perspective - adding something that's already there is safe because it doesn't change the result.

But I see you're thinking about it from the intent perspective - that we should never try to add something that's already there in the first place. This is a good safety check.

So I'd suggest keeping it as is, but can we find a better name for it?

Reply:
I'm definitely open to a better name -- do you have suggestions? `Add` is obviously taken. We could make it `AddAllowDuplicates`, or maybe something like `AddStructural` (not sure I love the second one myself).